saleor/discount/utils.py

```diff
@@ -56,14 +56,16 @@ def get_product_discount_on_sale(
     product_collections: Set[int],
     discount: DiscountInfo,
     channel: "Channel",
+    variant_id: Optional[int] = None,
 ):
     """Return discount value if product is on sale or raise NotApplicable."""
     is_product_on_sale = (
         product.id in discount.product_ids
         or product.category_id in discount.category_ids
         or product_collections.intersection(discount.collection_ids)
     )
-    if is_product_on_sale:
+    is_variant_on_sale = variant_id is not None and variant_id in discount.variants_ids
```

Comment:
Maybe we could simplify this?
`variant_id and variant_id in discount.variants_ids`

Reply:
Done, I just have a habit (maybe not the best one) to explicitly ask for `None` in order to avoid unwanted falsy values.